image/manifest_list.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2016 The Linux Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package image
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/opencontainers/image-spec/schema"
+	"github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+)
+
+type manifestlist struct {
+	Manifests []ManifestDescriptor `json:"manifests"`
+}
+type ManifestDescriptor v1.ManifestDescriptor
+
+func findManifestList(w walker, d *descriptor) (*manifestlist, error) {
+	var ml manifestlist
+	mlpath := filepath.Join("blobs", d.algo(), d.hash())
+
+	switch err := w.walk(func(path string, info os.FileInfo, r io.Reader) error {
+		if info.IsDir() || filepath.Clean(path) != mlpath {
+			return nil
+		}
+
+		buf, err := ioutil.ReadAll(r)
+		if err != nil {
+			return errors.Wrapf(err, "%s: error reading manifestlist", path)
+		}
+
+		if err := schema.MediaTypeManifestList.Validate(bytes.NewReader(buf)); err != nil {
+			return errors.Wrapf(err, "%s: manifestlist validation failed", path)
+		}
+
+		if err := json.Unmarshal(buf, &ml); err != nil {
+			return err
+		}
+
+		if len(ml.Manifests) == 0 {
+			return fmt.Errorf("%s: no manifests found", path)
+		}
+
+		return errEOW
+	}); err {
+	case nil:
+		return nil, fmt.Errorf("%s: manifestlist not found", mlpath)
+	case errEOW:
+		return &ml, nil
+	default:
+		return nil, err
+	}
+}
+
+func (ml *manifestlist) validate(w walker) error {
+	for _, d := range ml.Manifests {
+		if err := d.Descriptor.validate(w, []string{v1.MediaTypeImageConfig}); err != nil {
```

Comment:
But the manifest also have  [spec's Descriptor](https://github.com/opencontainers/image-spec/blob/master/specs-go/v1/manifest.go#L25), it can use [validate](https://github.com/opencontainers/image-tools/blob/master/image/manifest.go#L78) ,I think as long as the manifest file to the Descriptor as specific, you can call.